File: src/slack/knowledge.ts
import fs from 'fs-extra';
import path from 'path';
import { WikiUpdateItem, WikiSearchResult, SharedSourceItem, StagedQueueItem } from '../types.js';

export class ScopedKnowledgeBase {
  public rootDir: string;
  public sharedDir: string;
  public wikiDir: string;
  public sourcesDir: string;

  constructor(rootDir: string) {
    this.rootDir = path.resolve(rootDir);
    this.sharedDir = path.join(this.rootDir, 'data', 'shared');
    this.wikiDir = path.join(this.sharedDir, 'wiki');
    this.sourcesDir = path.join(this.sharedDir, 'sources');

    fs.ensureDirSync(this.wikiDir);
    fs.ensureDirSync(this.sourcesDir);
  }

  private assertSafePath(targetPath: string): string {
    const resolved = path.resolve(targetPath);
    const rel = path.relative(this.sharedDir, resolved);
    if (rel.startsWith('..') || path.isAbsolute(rel)) {
      throw new Error(`Access denied: Path '${targetPath}' is outside the shared knowledge boundary.`);
    }
    return resolved;
  }

  private getAllMarkdownFiles(dir: string): string[] {
    const files: string[] = [];
    if (!fs.existsSync(dir)) return files;
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        files.push(...this.getAllMarkdownFiles(fullPath));
      } else if (entry.isFile() && entry.name.endsWith('.md')) {
        files.push(fullPath);
      }
    }
    return files;
  }

  public getRecentWikiUpdates(days: number = 7): WikiUpdateItem[] {
    const cutoffTime = Date.now() - days * 24 * 60 * 60 * 1000;
    const recentNotes: WikiUpdateItem[] = [];

    const mdFiles = this.getAllMarkdownFiles(this.wikiDir);
    for (const mdFile of mdFiles) {
      try {
        const stat = fs.statSync(mdFile);
        if (stat.mtimeMs >= cutoffTime) {
          const relPath = path.relative(this.wikiDir, mdFile).replace(/\\/g, '/');
          const content = fs.readFileSync(mdFile, 'utf-8');

          let title = path.basename(mdFile, '.md');
          let summary = '';
          const tags: string[] = [];

          const lines = content.split('\n');
          for (const line of lines) {
            if (line.startsWith('# ') && title === path.basename(mdFile, '.md')) {
              title = line.replace(/^#\s+/, '').trim();
            } else if (line.toLowerCase().startsWith('tags:')) {
              const tagsPart = line.split(':', 2)[1] || '';
              const parts = tagsPart.split(/[,[\]]/).map((t) => t.trim()).filter(Boolean);
              tags.push(...parts);
            }
          }

          const nonHeaderLines = lines
            .map((l) => l.trim())
            .filter((l) => l && !l.startsWith('#') && !l.startsWith('---'));

          if (nonHeaderLines.length > 0) {
            summary = nonHeaderLines[0].slice(0, 200);
          }

          recentNotes.push({
            path: `wiki/${relPath}`,
            title,
            modified_at: new Date(stat.mtimeMs).toISOString(),
            summary,
            tags,
          });
        }
      } catch {}
    }

    recentNotes.sort((a, b) => new Date(b.modified_at).getTime() - new Date(a.modified_at).getTime());
    return recentNotes;
  }

  public searchSharedWiki(query: string, maxResults: number = 10): WikiSearchResult[] {
    const results: WikiSearchResult[] = [];
    if (!query.trim() || !fs.existsSync(this.wikiDir)) return results;

    const rawTokens = query.split(/\s+/).filter((t) => t.length > 1);
    const tokens = rawTokens.length > 0 ? rawTokens : [query.trim()];
    const escaped = tokens.map((t) => t.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|');
    const regex = new RegExp(escaped, 'gi');

    const mdFiles = this.getAllMarkdownFiles(this.wikiDir);
    for (const mdFile of mdFiles) {
      try {
        const content = fs.readFileSync(mdFile, 'utf-8');
        const matches = content.match(regex);
        if (matches && matches.length > 0) {
          const relPath = path.relative(this.wikiDir, mdFile).replace(/\\/g, '/');
          const score = matches.length;

          let title = path.basename(mdFile, '.md');
          for (const line of content.split('\n')) {
            if (line.startsWith('# ')) {
              title = line.replace(/^#\s+/, '').trim();
              break;
            }
          }

          let excerpt = '';
          const matchIndex = content.toLowerCase().indexOf(tokens[0].toLowerCase());
          if (matchIndex >= 0) {
            const start = Math.max(0, matchIndex - 100);
            const end = Math.min(content.length, matchIndex + 150);
            excerpt = `...${content.slice(start, end).replace(/\n/g, ' ')}...`;
          }

          results.push({
            path: `wiki/${relPath}`,
            title,
            score,
            excerpt,
          });
        }
      } catch {}
    }

    results.sort((a, b) => b.score - a.score);
    return results.slice(0, maxResults);
  }

  public readSharedNote(relativePath: string, maxChars: number = 6000): Record<string, any> {
    const cleanRel = relativePath.replace(/^\/+/, '').trim();
    const targetPath = path.join(this.sharedDir, cleanRel);

    let safePath: string;
    try {
      safePath = this.assertSafePath(targetPath);
    } catch (err: any) {
      return { error: err.message };
    }

    if (!fs.existsSync(safePath) || !fs.statSync(safePath).isFile()) {
      return { error: `File '${cleanRel}' not found in shared knowledge base.` };
    }

    try {
      const content = fs.readFileSync(safePath, 'utf-8');
      const isTruncated = content.length > maxChars;
      return {
        path: cleanRel,
        content: content.slice(0, maxChars),
        truncated: isTruncated,
        total_chars: content.length,
      };
    } catch (err: any) {
      return { error: `Failed to read file: ${err.message}` };
    }
  }

  public listSharedSources(limit: number = 20): { sources: SharedSourceItem[]; staged_queue: StagedQueueItem[] } {
    const sources: SharedSourceItem[] = [];
    const stagingFile = path.join(this.sourcesDir, 'staging_queue.json');
    let stagedItems: StagedQueueItem[] = [];

    if (fs.existsSync(stagingFile)) {
      try {
        stagedItems = fs.readJsonSync(stagingFile);
      } catch {
        stagedItems = [];
      }
    }

    if (fs.existsSync(this.sourcesDir)) {
      const entries = fs.readdirSync(this.sourcesDir, { withFileTypes: true });
      for (const entry of entries) {
        if (entry.name === 'staging_queue.json' || entry.name.startsWith('.')) continue;
        const fullPath = path.join(this.sourcesDir, entry.name);
        const stat = fs.statSync(fullPath);
        sources.push({
          name: entry.name,
          is_dir: entry.isDirectory(),
          size_bytes: entry.isFile() ? stat.size : null,
          modified_at: new Date(stat.mtimeMs).toISOString(),
        });
      }
    }

    sources.sort((a, b) => new Date(b.modified_at).getTime() - new Date(a.modified_at).getTime());
    return {
      sources: sources.slice(0, limit),
      staged_queue: stagedItems.slice(0, limit),
    };
  }

  public stageSourceUrl(
    url: string,
    title?: string,
    submitter: string = 'slack-user',
    notes: string = ''
  ): { status: string; message: string; entry: StagedQueueItem } {
    const stagingFile = path.join(this.sourcesDir, 'staging_queue.json');
    let queue: StagedQueueItem[] = [];
    if (fs.existsSync(stagingFile)) {
      try {
        queue = fs.readJsonSync(stagingFile);
      } catch {
        queue = [];
      }
    }

    const entry: StagedQueueItem = {
      url,
      title: title || url,
      submitted_by: submitter,
      notes,
      status: 'pending',
      staged_at: new Date().toISOString(),
    };

    queue.push(entry);
    fs.writeJsonSync(stagingFile, queue, { spaces: 2 });
    return {
      status: 'success',
      message: `Staged paper '${entry.title}' for ingestion.`,
      entry,
    };
  }
}

File: podarcislab/slack/knowledge.py
"""Scoped shared research knowledge base reader and staging tools for Slack agent.

STRICT PRIVACY ENCLOSURE:
This module only ever interacts with data/shared/ (wiki and sources).
It explicitly blocks access to data/users/ to protect individual researcher workspaces.
"""

import datetime
import json
import os
import re
import subprocess
from pathlib import Path
from typing import Any, Dict, List, Optional


class ScopedKnowledgeBase:
    """Provides safe, scoped access to shared research wiki and sources."""

    def __init__(self, root_dir: Path):
        self.root_dir = root_dir.resolve()
        self.shared_dir = (self.root_dir / "data" / "shared").resolve()
        self.wiki_dir = (self.shared_dir / "wiki").resolve()
        self.sources_dir = (self.shared_dir / "sources").resolve()

        # Ensure directories exist
        self.wiki_dir.mkdir(parents=True, exist_ok=True)
        self.sources_dir.mkdir(parents=True, exist_ok=True)

    def _assert_safe_path(self, target_path: Path) -> Path:
        """Validate that target_path is strictly within data/shared/."""
        resolved = target_path.resolve()
        try:
            resolved.relative_to(self.shared_dir)
        except ValueError:
            raise PermissionError(
                f"Access denied: Path '{target_path}' is outside the shared knowledge boundary."
            )
        return resolved

    def get_recent_wiki_updates(self, days: int = 7) -> List[Dict[str, Any]]:
        """Find wiki notes created or modified within the last N days."""
        cutoff_time = datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(days=days)
        recent_notes: List[Dict[str, Any]] = []

        if not self.wiki_dir.exists():
            return recent_notes

        # Scan all markdown files in shared wiki
        for md_file in self.wiki_dir.rglob("*.md"):
            try:
                stat = md_file.stat()
                mtime = datetime.datetime.fromtimestamp(stat.st_mtime, tz=datetime.timezone.utc)
                if mtime >= cutoff_time:
                    rel_path = md_file.relative_to(self.wiki_dir).as_posix()
                    content = md_file.read_text(encoding="utf-8", errors="replace")
                    
                    title = md_file.stem
                    summary = ""
                    tags = []

                    # Extract title and frontmatter if present
                    lines = content.splitlines()
                    for line in lines:
                        if line.startswith("# ") and title == md_file.stem:
                            title = line.lstrip("# ").strip()
                        elif line.startswith("tags:") or line.startswith("Tags:"):
                            tags_part = line.split(":", 1)[1]
                            tags = [t.strip() for t in re.split(r"[,\[\]]", tags_part) if t.strip()]

                    # Get brief excerpt
                    non_header_lines = [
                        l.strip()
                        for l in lines
                        if l.strip() and not l.startswith("#") and not l.startswith("---")
                    ]
                    if non_header_lines:
                        summary = non_header_lines[0][:200]

                    recent_notes.append(
                        {
                            "path": f"wiki/{rel_path}",
                            "title": title,
                            "modified_at": mtime.isoformat(),
                            "summary": summary,
                            "tags": tags,
                        }
                    )
            except Exception:
                continue

        recent_notes.sort(key=lambda x: x["modified_at"], reverse=True)
        return recent_notes

    def search_shared_wiki(self, query: str, max_results: int = 10) -> List[Dict[str, Any]]:
        """Search shared wiki notes for matching keywords or phrases."""
        results: List[Dict[str, Any]] = []
        if not self.wiki_dir.exists() or not query.strip():
            return results

        tokens = [re.escape(t.lower()) for t in query.split() if len(t) > 1]
        if not tokens:
            tokens = [re.escape(query.lower())]
        pattern = re.compile(r"|".join(tokens), re.IGNORECASE)

        for md_file in self.wiki_dir.rglob("*.md"):
            try:
                content = md_file.read_text(encoding="utf-8", errors="replace")
                matches = pattern.findall(content)
                if matches:
                    rel_path = md_file.relative_to(self.wiki_dir).as_posix()
                    score = len(matches)
                    
                    # Extract title
                    title = md_file.stem
                    for line in content.splitlines():
                        if line.startswith("# "):
                            title = line.lstrip("# ").strip()
                            break

                    # Excerpt surrounding first match
                    first_match = pattern.search(content)
                    excerpt = ""
                    if first_match:
                        start = max(0, first_match.start() - 100)
                        end = min(len(content), first_match.end() + 150)
                        excerpt = "..." + content[start:end].replace("\n", " ") + "..."

                    results.append(
                        {
                            "path": f"wiki/{rel_path}",
                            "title": title,
                            "score": score,
                            "excerpt": excerpt,
                        }
                    )
            except Exception:
                continue

        results.sort(key=lambda x: x["score"], reverse=True)
        return results[:max_results]

    def read_shared_note(self, relative_path: str, max_chars: int = 6000) -> Dict[str, Any]:
        """Read content of a note inside shared/wiki/ or shared/sources/."""
        # Strip leading slashes
        clean_rel = relative_path.lstrip("/").strip()
        target_path = self.shared_dir / clean_rel
        safe_path = self._assert_safe_path(target_path)

        if not safe_path.exists() or not safe_path.is_file():
            return {"error": f"File '{clean_rel}' not found in shared knowledge base."}

        try:
            content = safe_path.read_text(encoding="utf-8", errors="replace")
            is_truncated = len(content) > max_chars
            return {
                "path": clean_rel,
                "content": content[:max_chars],
                "truncated": is_truncated,
                "total_chars": len(content),
            }
        except Exception as e:
            return {"error": f"Failed to read file: {str(e)}"}

    def list_shared_sources(self, limit: int = 20) -> Dict[str, Any]:
        """List literature files and staged papers in data/shared/sources/."""
        sources = []
        staging_file = self.sources_dir / "staging_queue.json"
        staged_items = []

        if staging_file.exists():
            try:
                staged_items = json.loads(staging_file.read_text(encoding="utf-8"))
            except Exception:
                staged_items = []

        for item in self.sources_dir.iterdir():
            if item.name == "staging_queue.json" or item.name.startswith("."):
                continue
            stat = item.stat()
            sources.append(
                {
                    "name": item.name,
                    "is_dir": item.is_dir(),
                    "size_bytes": stat.st_size if item.is_file() else None,
                    "modified_at": datetime.datetime.fromtimestamp(
                        stat.st_mtime, tz=datetime.timezone.utc
                    ).isoformat(),
                }
            )

        sources.sort(key=lambda x: x["modified_at"], reverse=True)
        return {
            "sources": sources[:limit],
            "staged_queue": staged_items[:limit],
        }

    def stage_source_url(
        self, url: str, title: Optional[str] = None, submitter: str = "slack-user", notes: str = ""
    ) -> Dict[str, Any]:
        """Stage a paper or URL into data/shared/sources/staging_queue.json for team ingestion."""
        staging_file = self.sources_dir / "staging_queue.json"
        queue = []
        if staging_file.exists():
            try:
                queue = json.loads(staging_file.read_text(encoding="utf-8"))
            except Exception:
                queue = []

        entry = {
            "url": url,
            "title": title or url,
            "submitted_by": submitter,
            "notes": notes,
            "status": "pending",
            "staged_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        }
        queue.append(entry)
        staging_file.write_text(json.dumps(queue, indent=2), encoding="utf-8")
        return {"status": "success", "message": f"Staged paper '{entry['title']}' for ingestion.", "entry": entry}
